### PR TITLE
#5 Query tracing

### PR DESCRIPTION
**Pull request by MichiruKayo**
Output example

```
2017-11-02 16:42:07.530  INFO 14368 --- [  XNIO-2 task-1] GraphqlTracingInstrumentation            : Query (example) Execution time : 56ms
2017-11-02 16:42:07.530  INFO 14368 --- [  XNIO-2 task-1] GraphqlTracingInstrumentation            : Query (example) : query example($id:String){
  ping
  jokes{
    jokeByCategory(request: {category: NERDY}){
      id
      text
    }
    jokeById(id: $id){
      id
      text
    }
  }
}

2017-11-02 16:42:07.530  INFO 14368 --- [  XNIO-2 task-1] GraphqlTracingInstrumentation            : Query (example) Variables : {id=1}

```

### CODE REVIEW

**Comment by MichiruKayo** [comment]
Yes you are right.. 
If you use default Execution Strategy (async), then there is such row: 
log.debug("Executing request. operation name: {}. query: {}. variables {} ", executionInput.getOperationName(), executionInput.getQuery(), executionInput.getVariables());
(and this is the only debug message they have) 

But it does not print execution time, to do this you need to enable TracingInstrumentation (which tracks and stores all this information ) and print it manually.

**Comment by MichiruKayo** [comment]
i have changed instrumentation to print only execution time, and updated read.me to enable debug level.

File: api/src/main/java/lv/ctco/tpl/bff/configuration/GraphqlTracingInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+@Component
+public class GraphqlTracingInstrumentation extends TracingInstrumentation {
+
+    private static final String DURATION = "duration";
+    private final Logger log = Logger.getLogger(GraphqlTracingInstrumentation.class.getSimpleName());
```
Comment (trioletas) [inline api/src/main/java/lv/ctco/tpl/bff/configuration/GraphqlTracingInstrumentation.java]: please use lombok's @Slf4j

File: api/src/main/java/lv/ctco/tpl/bff/configuration/GraphqlTracingInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Override
+    public CompletableFuture<ExecutionResult> instrumentExecutionResult(ExecutionResult executionResult,
+                                                                        InstrumentationExecutionParameters parameters) {
+        TracingSupport tracingSupport = parameters.getInstrumentationState();
+        Long executionTime = getTime(tracingSupport.snapshotTracingData());
+        log.info(String.format("Query (%s) Execution time : %sms", parameters.getOperation(), executionTime));
```
Comment (trioletas) [inline api/src/main/java/lv/ctco/tpl/bff/configuration/GraphqlTracingInstrumentation.java]: `executionTime` is nullable -- does it make sense to log if if the result is null?
`"Query (bla) Execution time : nullms`

Comment (trioletas) [inline api/src/main/java/lv/ctco/tpl/bff/configuration/GraphqlTracingInstrumentation.java]: for readability, please add a white space between the time and the `ms` suffix -> `%s ms`

Comment (trioletas) [inline api/src/main/java/lv/ctco/tpl/bff/configuration/GraphqlTracingInstrumentation.java]: please downgrade to `debug` to stay consistent with default tracing level
the usual
```java
if log.isDebugEnabled
....
```

File: api/src/main/java/lv/ctco/tpl/bff/configuration/GraphqlTracingInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import graphql.execution.instrumentation.tracing.TracingSupport;
+import org.springframework.stereotype.Component;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+@Component
+public class GraphqlTracingInstrumentation extends TracingInstrumentation {
```
Comment (trioletas) [inline api/src/main/java/lv/ctco/tpl/bff/configuration/GraphqlTracingInstrumentation.java]: please rename to `GraphqlTracingInstrumentationConfiguration` to stay consistent with our files in that package

**Comment by MichiruKayo** [comment]
Made code review changes. also added "topic=graphql" to sl4j for logging consistency

**Comment by trioletas** [comment]
Awesome, thanks!

**Pull request merged**